### ISSUE TITLE
Inconsistency path mapping option

### ISSUE THREAD

**Issue by JVApen** [clangd/clangd#2100]
We make an index on a dedicated machine to be used later by many users. The 'working directory' of that machine is different than the one of the user. To make this index usable, the --path-mappings option can be used to map between those 2 paths and make it usable.

Indicating which (file) index is used can be done by specifying it in the .clangd file or by specifying it on the command line. However specifying the file-mappings can only be done via the command line.

As we (want to) use a vs code startup task to select the right index and create/update the .clangd, this is annoying as we now also have to update the vs code settings in order to change the clangd command line.

**System information**

Output of `clangd --version`: 18.1.3

Editor/LSP plugin: VS Code + standard clangd plugin

Operating system: Windows

**Comment by HighCommander4** [comment]
> We make an index on a dedicated machine to be used later by many users. The 'working directory' of that machine is different than the one of the user. To make this index usable, the --path-mappings option can be used to map between those 2 paths and make it usable.

I don't think `--path-mappings` is suitable for this purpose.

I believe `--path-mappings` is intended for cases where the client (editor) and server (clangd) run on different machines.

If you're running clangd locally, and using `--path-mappings`, the clangd will use the mapped paths for everything (not only index operations, but e.g. when building the AST for an open file and it needs to read headers from disk). I think that's likely to break things.

**Comment by JVApen** [comment]
@HighCommander4  Strangely enough, it seems to be working quite nice if I use `--path-mappings=E:/checkout/=E:/generation/directory` (location of slashes is important) It might not be the intention, though as far as I see, this is the only way that we can make this mapping happen.
Thanks for the warning, I'll ask for some extra testing to see if nothing is broken.

**Comment by HighCommander4** [comment]
> as far as I see, this is the only way that we can make this mapping happen

I looked around a bit to try to figure out what is the intended way to address the problem of mapping from local paths to paths on the machine where the index file was created.

It looks like, at least in the case where the generated index is served by `clangd-index-server` (used for clangd's [remote index](https://clangd.llvm.org/guides/remote-index) feature, though it can also be run locally), the idea is to use the `--project-root` flag of `clangd-index-server` in combination with `MountPoint:`. So in your case, you would:
 * run `clangd-index-server` with `--project-root=E:/generation/directory`
 * consume the index with `MountPoint: E:/checkout/`

However, while `MountPoint:` applies to the case of consuming a index file directly (without `clangd-index-server`), I couldn't find an equivalent of `--projec-root` to strip of the path prefix on the machine where the index file was generated.

It's possible that this scenario was overlooked; my understanding is that `clangd-indexer` is most commonly used in combination with `clangd-index-server`.

**Comment by JVApen** [comment]
I understand that the index server would be used more. Though it also comes with restrictions. In our setup (where we had a custom clangd executable that we are replacing/upgrading) we have an index file based on the git revision the file is created on. So 2 users can be using a different index at the same time. We also have users looking at multiple releases of the code without closing their IDE. Trying to get this done with the index server is quite some complexity. One would already have to start an index server together with the clangd and use that one. I'm not saying that is impossible, though it's a big step in the upgrade.

Am I correct that if Spec.MountPoint would be filled in here, we would expect it to work? https://github.com/llvm/llvm-project/blob/f4be6812e203690073280b9ac8d60092d75bbdce/clang-tools-extra/clangd/tool/ClangdMain.cpp#L679

**Comment by HighCommander4** [comment]
> Am I correct that if Spec.MountPoint would be filled in here, we would expect it to work? https://github.com/llvm/llvm-project/blob/f4be6812e203690073280b9ac8d60092d75bbdce/clang-tools-extra/clangd/tool/ClangdMain.cpp#L679

I don't expect setting `Spec.MountPoint` there to have any effect. As far as I can tell, the only place `MountPoint` is used in the `Kind == File` case is [here](https://searchfox.org/llvm/rev/f916cb6184fb04b27ae7b867edcfd162c3a49694/clang-tools-extra/clangd/ConfigCompile.cpp#403) (to filter which files the index applies to), and `FlagsConfigProvider` bypasses that (since command line flags apply to all files).

**Comment by JVApen** [comment]
Indeed, I did some tracing and for some point the MountPoint and the ProjectRoot are considered one variable. However, than it called remote::getClient() which creates a Marshaller. I believe that this is where the actual path mapping is happening between the file paths of index creation and usage. This making use of `replace_path_prefix(RelativePath, LocalIndexRoot, ""))` and `replace_path_prefix(Result, RemoteIndexRoot, "")`
